test/functional/terminal/ex_terminal_spec.lua

```diff
@@ -48,6 +49,39 @@ describe(':terminal', function()
     eq(2, eval("line('.')"))  -- cursor stays where we put it
   end)
 
+  it('does not crash when callback fires after terminal closes', function()
+    screen:detach()
+    screen = thelpers.screen_setup(0, '["'..helpers.nvim_prog..'", "-u", "NONE", "-i", "NONE", "-N"]')
+    thelpers.feed_data({':let g:split = 0',
+                        ':au TermOpen * nnoremap <buffer> x :q<CR>',
+                        ':nnoremap <silent> x :new <bar> let g:split += 1 <bar> call termopen([&shell])<CR>',
+                        ':set updatetime=0',
+                        ''})
+    screen:expect([[
+      {1: }                                                 |
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {5:[No Name]                                         }|
+      :set updatetime=0                                 |
+      {3:-- TERMINAL --}                                    |
+    ]])
+    local count = 500
+    for _ = 1,count do
+      thelpers.feed_data('xx')
+      wait()
+    end
+    thelpers.feed_data({':echo g:split', ''})
+    screen:expect([[
+      {1: }                                                 |
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {4:~                                                 }|
+      {5:[No Name]                                         }|
+      ]]..tostring(count)..string.rep(' ', 50 - string.len(tostring(count)))..[[|
+      {3:-- TERMINAL --}                                    |
```

Comment:
We mostly just want to give ASAN a chance. So maybe 10 iterations instead of 500 is good enough.

All of the screen redraws will cause screen:expect() to be really slow, and cause too many "indetermistic" screen states. Some of the envs on travis are choking. 

We only really care about waiting for the child nvim to process the data, so perhaps try screen:expect() with the "any" parameter:

    screen:expect(tostring(count)..string.rep(' ', 50 - string.len(tostring(count))), 
      nil, nil, nil, true)